File: golang/serve.go
// A2A (agent-to-agent) — INBOUND: serve a toolkit as a remote A2A agent.
//
// Toolkit.Serve(addr, ServeOptions{Client, A2A}) stands up a minimal HTTP server
// that, when the A2A profile is present, exposes:
//
//	GET  /.well-known/agent-card.json  → an Agent Card built from the toolkit's
//	     SKILLS (SKILL.md name+description), never raw tools.
//	POST /                             → JSON-RPC 2.0: SendMessage (submit a Task,
//	     fulfil it asynchronously via Client.Run) + GetTask (poll).
//
// This is the inbound counterpart to a2a.go (outbound) and speaks the same wire
// subset (A2ATask/A2AArtifact/A2AMessage/A2APart are reused verbatim). Task
// persistence is a pluggable TaskStore (in-memory default; file / custom
// selectable via A2AConfig.Store). See ../SPEC.md §7B and
// openspec/changes/add-a2a-agents.
package toolnexus

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

const (
	cardPath             = "/.well-known/agent-card.json"
	serveProtocolVersion = "0.3.0"
	serveDefaultVersion  = "0.1.0"
)

// ---------------------------------------------------------------------------
// TaskStore adapter — pluggable persistence for served Tasks.
// ---------------------------------------------------------------------------

// TaskStore is pluggable Task persistence. A nil *A2ATask from Get signals an
// unknown id (the JS `undefined`); Get returns an error only on an unexpected
// failure a custom store wants to surface. The-factory can plug NATS/JetStream
// the same way.
type TaskStore interface {
	Get(id string) (*A2ATask, error)
	Save(task A2ATask) error
}

// InMemoryTaskStore is the default store — Tasks live only for the process
// lifetime. Safe for concurrent use.
type InMemoryTaskStore struct {
	mu    sync.Mutex
	tasks map[string]A2ATask
}

// NewInMemoryTaskStore builds an empty in-memory store.
func NewInMemoryTaskStore() *InMemoryTaskStore {
	return &InMemoryTaskStore{tasks: map[string]A2ATask{}}
}

// Get returns the Task by id, or (nil, nil) when absent.
func (s *InMemoryTaskStore) Get(id string) (*A2ATask, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	t, ok := s.tasks[id]
	if !ok {
		return nil, nil
	}
	clone := t
	return &clone, nil
}

// Save persists the Task by its id.
func (s *InMemoryTaskStore) Save(task A2ATask) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.tasks[task.ID] = task
	return nil
}

// FileTaskStore is a file-backed store — one JSON file per Task id, under dir.
type FileTaskStore struct {
	dir string
}

// NewFileTaskStore builds a file store rooted at dir (created if missing).
func NewFileTaskStore(dir string) *FileTaskStore {
	_ = os.MkdirAll(dir, 0o755)
	return &FileTaskStore{dir: dir}
}

func (s *FileTaskStore) file(id string) string {
	return filepath.Join(s.dir, Sanitize(id)+".json")
}

// Get reads the Task from its file, or (nil, nil) when the file is missing or
// unreadable (mirrors the JS store, which swallows read errors as undefined).
func (s *FileTaskStore) Get(id string) (*A2ATask, error) {
	data, err := os.ReadFile(s.file(id))
	if err != nil {
		return nil, nil
	}
	var t A2ATask
	if err := json.Unmarshal(data, &t); err != nil {
		return nil, nil
	}
	return &t, nil
}

// Save writes the Task to its file.
func (s *FileTaskStore) Save(task A2ATask) error {
	data, err := json.Marshal(task)
	if err != nil {
		return err
	}
	return os.WriteFile(s.file(task.ID), data, 0o644)
}

// ResolveStore maps a store selector to a concrete TaskStore:
// nil / "" / "memory" ⇒ in-memory; "file:<dir>" ⇒ file store; a TaskStore value
// ⇒ used as-is. Any other string is an error.
func ResolveStore(store any) (TaskStore, error) {
	switch v := store.(type) {
	case nil:
		return NewInMemoryTaskStore(), nil
	case TaskStore:
		return v, nil
	case string:
		if v == "" || v == "memory" {
			return NewInMemoryTaskStore(), nil
		}
		if strings.HasPrefix(v, "file:") {
			return NewFileTaskStore(strings.TrimPrefix(v, "file:")), nil
		}
		return nil, fmt.Errorf("unknown A2A store: %q (expected \"memory\" or \"file:<dir>\")", v)
	default:
		return nil, fmt.Errorf("unknown A2A store type: %T", store)
	}
}

// ---------------------------------------------------------------------------
// A2A serve config + handle.
// ---------------------------------------------------------------------------

// A2AProvider identifies the organization publishing the agent.
type A2AProvider struct {
	Organization string `json:"organization"`
	URL          string `json:"url"`
}

// A2AConfig is the opt-in A2A profile for Serve — it configures the Agent Card
// and the Task store.
type A2AConfig struct {
	Name        string       `json:"name,omitempty"`
	Description string       `json:"description,omitempty"`
	Version     string       `json:"version,omitempty"`
	Provider    *A2AProvider `json:"provider,omitempty"`
	// Skills is the subset of the toolkit's skill names to advertise; nil ⇒ all,
	// non-nil (even empty) ⇒ filter to exactly this set.
	Skills []string `json:"skills,omitempty"`
	// Store is "memory" (default) | "file:<dir>" | a custom TaskStore.
	Store any `json:"store,omitempty"`
}

// OnTaskEvent is surfaced on each Task's terminal state, carrying the RunResult
// telemetry.
type OnTaskEvent struct {
	ID     string
	Skill  string
	Task   A2ATask
	Result *RunResult
	State  string
}

// OnTask is the terminal-state callback passed to Serve.
type OnTask func(ev OnTaskEvent)

// ServeOptions configures Toolkit.Serve. Client fulfils Tasks via its run loop;
// A2A (or the toolkit's top-level a2a config block) enables the A2A routes.
type ServeOptions struct {
	Client *Client
	A2A    *A2AConfig
	OnTask OnTask
}

// ServeHandle is returned by Serve — the base URL plus a stop/close method.
type ServeHandle struct {
	// URL is the base URL of the server, e.g. http://127.0.0.1:PORT.
	URL string

	server *http.Server
	once   sync.Once
	err    error
}

// Stop shuts the server down (idempotent).
func (h *ServeHandle) Stop() error {
	h.once.Do(func() {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		h.err = h.server.Shutdown(ctx)
	})
	return h.err
}

// Close is an alias for Stop.
func (h *ServeHandle) Close() error { return h.Stop() }

// ---------------------------------------------------------------------------
// Card builder.
// ---------------------------------------------------------------------------

// agentCardCapabilities is the fixed capability block of a served card. A struct
// (not a map) so the JSON field order and empty values match the JS reference
// byte-for-byte.
type agentCardCapabilities struct {
	Streaming         bool `json:"streaming"`
	PushNotifications bool `json:"pushNotifications"`
}

// agentCardSkill is one advertised skill on a served card ({id, name, description}
// — always all three, even when description is empty).
type agentCardSkill struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
}

// servedAgentCard is the exact wire shape published at cardPath. A dedicated
// struct (rather than the outbound AgentCard, whose omitempty tags would drop
// description:"" and whose map capabilities would reorder keys) so the served
// card is byte-identical to the JS reference (SPEC §7B).
type servedAgentCard struct {
	Name               string                `json:"name"`
	Description        string                `json:"description"`
	Version            string                `json:"version"`
	ProtocolVersion    string                `json:"protocolVersion"`
	Capabilities       agentCardCapabilities `json:"capabilities"`
	DefaultInputModes  []string              `json:"defaultInputModes"`
	DefaultOutputModes []string              `json:"defaultOutputModes"`
	Skills             []agentCardSkill      `json:"skills"`
	URL                string                `json:"url"`
	Provider           *A2AProvider          `json:"provider,omitempty"`
}

// BuildAgentCard builds an Agent Card from the toolkit's skills. skills are the
// SkillSource's SkillInfos (never raw tools); filtered to cfg.Skills when given.
// url is the JSON-RPC endpoint peers POST to.
func BuildAgentCard(cfg *A2AConfig, skills []SkillInfo, url string) servedAgentCard {
	selected := skills
	if cfg != nil && cfg.Skills != nil {
		wanted := make(map[string]bool, len(cfg.Skills))
		for _, name := range cfg.Skills {
			wanted[name] = true
		}
		selected = selected[:0:0]
		for _, s := range skills {
			if wanted[s.Name] {
				selected = append(selected, s)
			}
		}
	}
	cardSkills := make([]agentCardSkill, 0, len(selected))
	for _, s := range selected {
		cardSkills = append(cardSkills, agentCardSkill{ID: s.Name, Name: s.Name, Description: s.Description})
	}

	name := "toolnexus-agent"
	description := ""
	version := serveDefaultVersion
	var provider *A2AProvider
	if cfg != nil {
		if cfg.Name != "" {
			name = cfg.Name
		}
		if cfg.Description != "" {
			description = cfg.Description
		}
		if cfg.Version != "" {
			version = cfg.Version
		}
		provider = cfg.Provider
	}

	return servedAgentCard{
		Name:               name,
		Description:        description,
		Version:            version,
		ProtocolVersion:    serveProtocolVersion,
		Capabilities:       agentCardCapabilities{Streaming: false, PushNotifications: false},
		DefaultInputModes:  []string{"text"},
		DefaultOutputModes: []string{"text"},
		Skills:             cardSkills,
		URL:                url,
		Provider:           provider,
	}
}

// ---------------------------------------------------------------------------
// Server.
// ---------------------------------------------------------------------------

// startServerOptions is the internal config for startA2AServer (Toolkit.Serve
// delegates here).
type startServerOptions struct {
	addr    string
	a2a     *A2AConfig
	skills  []SkillInfo
	runTask func(text string) (RunResult, error)
	onTask  OnTask
}

// rpcError is the JSON-RPC 2.0 error object.
type rpcError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// rpcResponse is a JSON-RPC 2.0 response envelope. A struct (not a map) so the
// field order — jsonrpc, id, then result|error — matches the JS reference.
type rpcResponse struct {
	JSONRPC string    `json:"jsonrpc"`
	ID      any       `json:"id"`
	Result  any       `json:"result,omitempty"`
	Error   *rpcError `json:"error,omitempty"`
}

// messageText extracts the concatenated text of a SendMessage message's parts.
func messageText(params json.RawMessage) string {
	var p struct {
		Message struct {
			Parts []struct {
				Kind string `json:"kind"`
				Text string `json:"text"`
			} `json:"parts"`
		} `json:"message"`
	}
	if err := json.Unmarshal(params, &p); err != nil {
		return ""
	}
	var b strings.Builder
	for _, part := range p.Message.Parts {
		if part.Kind == "text" {
			b.WriteString(part.Text)
		}
	}
	return b.String()
}

// startA2AServer starts the HTTP server. Delegated to by Toolkit.Serve. When a2a
// is absent, the server answers 404 to everything (a minimal base for now).
func startA2AServer(opts startServerOptions) (*ServeHandle, error) {
	host, port := splitAddr(opts.addr)

	var store TaskStore
	if opts.a2a != nil {
		s, err := ResolveStore(opts.a2a.Store)
		if err != nil {
			return nil, err
		}
		store = s
	}

	ln, err := net.Listen("tcp", net.JoinHostPort(host, port))
	if err != nil {
		return nil, err
	}
	actualPort := ln.Addr().(*net.TCPAddr).Port
	base := baseURL(host, actualPort)

	writeRPC := func(w http.ResponseWriter, resp rpcResponse) {
		resp.JSONRPC = "2.0"
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(resp)
	}
	rpcOK := func(w http.ResponseWriter, id any, result any) {
		writeRPC(w, rpcResponse{ID: id, Result: result})
	}
	rpcErr := func(w http.ResponseWriter, id any, code int, message string) {
		writeRPC(w, rpcResponse{ID: id, Error: &rpcError{Code: code, Message: message}})
	}

	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// No A2A profile ⇒ no routes.
		if opts.a2a == nil || store == nil {
			w.WriteHeader(http.StatusNotFound)
			_, _ = w.Write([]byte("not found"))
			return
		}

		if r.Method == http.MethodGet && r.URL.Path == cardPath {
			w.Header().Set("content-type", "application/json")
			_ = json.NewEncoder(w).Encode(BuildAgentCard(opts.a2a, opts.skills, base+"/"))
			return
		}

		if r.Method == http.MethodPost {
			body, _ := io.ReadAll(r.Body)
			var rpc struct {
				ID     any             `json:"id"`
				Method string          `json:"method"`
				Params json.RawMessage `json:"params"`
			}
			if err := json.Unmarshal(body, &rpc); err != nil {
				rpcErr(w, nil, -32700, "Parse error")
				return
			}
			switch rpc.Method {
			case "SendMessage":
				text := messageText(rpc.Params)
				id := uuid.NewString()
				submitted := A2ATask{ID: id, Status: A2ATaskStatus{State: "submitted"}}
				// Persist submitted, kick fulfilment async, return the id now.
				_ = store.Save(submitted)
				go fulfil(id, text, store, opts.runTask, opts.onTask)
				rpcOK(w, rpc.ID, submitted)
			case "GetTask":
				var p struct {
					ID string `json:"id"`
				}
				_ = json.Unmarshal(rpc.Params, &p)
				task, _ := store.Get(p.ID)
				if task == nil {
					rpcErr(w, rpc.ID, -32001, "Task not found: "+p.ID)
					return
				}
				rpcOK(w, rpc.ID, *task)
			default:
				rpcErr(w, rpc.ID, -32601, "Method not found: "+rpc.Method)
			}
			return
		}

		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte("not found"))
	})

	srv := &http.Server{Handler: handler}
	go func() { _ = srv.Serve(ln) }()

	return &ServeHandle{URL: base, server: srv}, nil
}

// fulfil runs background fulfilment: submitted → working → (completed | failed).
// It never crashes the server — a fulfilment error (or panic) becomes a failed
// Task. Mirrors the JS fulfil().
func fulfil(
	id, text string,
	store TaskStore,
	runTask func(text string) (RunResult, error),
	onTask OnTask,
) {
	// A panic in the client loop must never crash the process.
	defer func() { _ = recover() }()

	var task A2ATask
	var result *RunResult
	var state string

	_ = store.Save(A2ATask{ID: id, Status: A2ATaskStatus{State: "working"}})
	rr, err := runTask(text)
	if err != nil {
		detail := err.Error()
		task = A2ATask{
			ID: id,
			Status: A2ATaskStatus{
				State:   "failed",
				Message: &A2AMessage{Role: "agent", Parts: []A2APart{{Kind: "text", Text: &detail}}},
			},
		}
		state = "failed"
	} else {
		result = &rr
		artifactText := rr.Text
		task = A2ATask{
			ID:     id,
			Status: A2ATaskStatus{State: "completed"},
			Artifacts: []A2AArtifact{{
				ArtifactID: uuid.NewString(),
				Parts:      []A2APart{{Kind: "text", Text: &artifactText}},
			}},
		}
		state = "completed"
	}

	// A store write failure must not crash the server.
	_ = store.Save(task)

	if onTask != nil {
		// Host callback errors/panics are isolated.
		func() {
			defer func() { _ = recover() }()
			onTask(OnTaskEvent{ID: id, Task: task, Result: result, State: state})
		}()
	}
}

// ---------------------------------------------------------------------------
// Address helpers.
// ---------------------------------------------------------------------------

// splitAddr splits host:port on the last colon. Missing host ⇒ 127.0.0.1.
func splitAddr(addr string) (host, port string) {
	idx := strings.LastIndex(addr, ":")
	if idx == -1 {
		return "127.0.0.1", addr
	}
	host = addr[:idx]
	if host == "" {
		host = "127.0.0.1"
	}
	return host, addr[idx+1:]
}

// baseURL is the base URL for a listening server; 0.0.0.0 / :: is reported as
// 127.0.0.1 for callers.
func baseURL(host string, port int) string {
	h := host
	if h == "0.0.0.0" || h == "::" {
		h = "127.0.0.1"
	}
	return fmt.Sprintf("http://%s:%d", h, port)
}

File: js/src/serve.ts
/**
 * A2A (agent-to-agent) — INBOUND: serve a toolkit as a remote A2A agent.
 *
 * `toolkit.serve(addr, { client, a2a })` stands up a minimal HTTP server that,
 * when the `a2a` profile is present, exposes:
 *
 *   GET  /.well-known/agent-card.json  → an Agent Card built from the toolkit's
 *        **skills** (SKILL.md name+description), never raw tools.
 *   POST /                             → JSON-RPC 2.0: `SendMessage` (submit a
 *        Task, fulfil it asynchronously via `client.run`) + `GetTask` (poll).
 *
 * This is the inbound counterpart to a2a.ts (outbound) and speaks the same wire
 * subset (verified against a2a-python). Task persistence is a pluggable
 * `TaskStore` (in-memory default; file / custom selectable via `a2a.store`).
 * See ../../SPEC.md and openspec/changes/add-a2a-agents.
 */
import http from "node:http"
import fs from "node:fs"
import path from "node:path"
import { randomUUID } from "node:crypto"
import { StreamableHTTPServerTransport } from "@modelcontextprotocol/sdk/server/streamableHttp.js"
import { sanitize, type Tool } from "./types.js"
import type { SkillInfo } from "./skill.js"
import type { RunResult } from "./client.js"
import type { A2AArtifact, A2ATask, A2ATaskState, AgentCard } from "./a2a.js"
import { buildMcpServer, exposedMcpTools, type MCPServeConfig, type OnCall } from "./mcpserve.js"

// ---------------------------------------------------------------------------
// TaskStore adapter — pluggable persistence for served Tasks.
// ---------------------------------------------------------------------------

/** Pluggable Task persistence. The-factory can plug NATS/JetStream the same way. */
export interface TaskStore {
  get(id: string): Promise<A2ATask | undefined>
  save(task: A2ATask): Promise<void>
}

/** Default in-memory store — Tasks live only for the lifetime of the process. */
export class InMemoryTaskStore implements TaskStore {
  private readonly tasks = new Map<string, A2ATask>()
  async get(id: string): Promise<A2ATask | undefined> {
    return this.tasks.get(id)
  }
  async save(task: A2ATask): Promise<void> {
    this.tasks.set(task.id, task)
  }
}

/** File-backed store — one JSON file per Task id, under `dir`. */
export class FileTaskStore implements TaskStore {
  constructor(private readonly dir: string) {
    fs.mkdirSync(dir, { recursive: true })
  }
  private file(id: string): string {
    return path.join(this.dir, `${sanitize(id)}.json`)
  }
  async get(id: string): Promise<A2ATask | undefined> {
    try {
      return JSON.parse(fs.readFileSync(this.file(id), "utf8")) as A2ATask
    } catch {
      return undefined
    }
  }
  async save(task: A2ATask): Promise<void> {
    fs.writeFileSync(this.file(task.id), JSON.stringify(task))
  }
}

/** Map a `store` selector to a concrete TaskStore. `"file:<dir>"` → FileTaskStore. */
export function resolveStore(store?: TaskStore | "memory" | string): TaskStore {
  if (store === undefined || store === "memory") return new InMemoryTaskStore()
  if (typeof store === "string") {
    if (store.startsWith("file:")) return new FileTaskStore(store.slice("file:".length))
    throw new Error(`Unknown A2A store: "${store}" (expected "memory" or "file:<dir>")`)
  }
  return store
}

// ---------------------------------------------------------------------------
// A2A serve config + handle.
// ---------------------------------------------------------------------------

export interface A2AProvider {
  organization: string
  url: string
}

/** Opt-in A2A profile for `toolkit.serve` — configures the Agent Card + store. */
export interface A2AConfig {
  name?: string
  description?: string
  version?: string
  provider?: A2AProvider
  /** Subset of the toolkit's skill ids/names to advertise; omit ⇒ all. */
  skills?: string[]
  /** `"memory"` (default) | `"file:<dir>"` | a custom TaskStore. */
  store?: TaskStore | "memory" | string
}

/** Surfaced on each Task's terminal state, carrying the RunResult telemetry. */
export interface OnTaskEvent {
  id: string
  skill?: string
  task: A2ATask
  result?: RunResult
  state: A2ATaskState
}

export type OnTask = (ev: OnTaskEvent) => void | Promise<void>

/** Handle returned by `toolkit.serve` — the base URL plus a stop/close method. */
export interface ServeHandle {
  /** Base URL of the server, e.g. `http://127.0.0.1:PORT`. */
  url: string
  stop(): Promise<void>
  /** Alias for `stop()`. */
  close(): Promise<void>
}

// ---------------------------------------------------------------------------
// Card builder.
// ---------------------------------------------------------------------------

const PROTOCOL_VERSION = "0.3.0"
const DEFAULT_VERSION = "0.1.0"

/**
 * Build an Agent Card from the toolkit's skills. `skills` are the SkillSource's
 * `SkillInfo`s (never raw tools); filtered to `cfg.skills` when given. `url` is
 * the JSON-RPC endpoint peers POST to.
 */
export function buildAgentCard(cfg: A2AConfig, skills: SkillInfo[], url: string): AgentCard {
  const wanted = cfg.skills
  const selected = wanted ? skills.filter((s) => wanted.includes(s.name)) : skills
  const card: AgentCard = {
    name: cfg.name ?? "toolnexus-agent",
    description: cfg.description ?? "",
    version: cfg.version ?? DEFAULT_VERSION,
    protocolVersion: PROTOCOL_VERSION,
    capabilities: { streaming: false, pushNotifications: false },
    defaultInputModes: ["text"],
    defaultOutputModes: ["text"],
    skills: selected.map((s) => ({ id: s.name, name: s.name, description: s.description ?? "" })),
    url,
  }
  if (cfg.provider) card.provider = cfg.provider
  return card
}

// ---------------------------------------------------------------------------
// Server.
// ---------------------------------------------------------------------------

const CARD_PATH = "/.well-known/agent-card.json"

export interface StartServerOptions {
  /** `host:port` (port 0 ⇒ ephemeral). */
  addr: string
  /** When present, mounts the A2A card + JSON-RPC routes; absent ⇒ no A2A routes. */
  a2a?: A2AConfig
  /** The toolkit's skills, for the card. */
  skills: SkillInfo[]
  /** Runs one served Task through the client loop. `contextId` (from the A2A
   * message) keys the conversation so served turns are remembered via the store. */
  runTask: (text: string, contextId?: string) => Promise<RunResult>
  onTask?: OnTask
  /** When present, mounts a streamable-HTTP MCP server at `/mcp`; absent ⇒ no MCP route. */
  mcp?: MCPServeConfig
  /** The toolkit's unified tools, exposed by the MCP profile. */
  mcpTools?: Tool[]
  /** Surfaced on each inbound MCP `tools/call`. */
  onCall?: OnCall
}

/** Extract the concatenated text of a JSON-RPC `SendMessage` message's parts. */
function messageText(params: any): string {
  const parts = params?.message?.parts
  if (!Array.isArray(parts)) return ""
  return parts
    .filter((p: any) => p && p.kind === "text" && typeof p.text === "string")
    .map((p: any) => p.text)
    .join("")
}

function readBody(req: http.IncomingMessage): Promise<string> {
  return new Promise((resolve) => {
    let body = ""
    req.on("data", (c) => (body += c))
    req.on("end", () => resolve(body))
  })
}

function safeJson(body: string): unknown {
  try {
    return JSON.parse(body)
  } catch {
    return undefined
  }
}

/**
 * Start the HTTP server. Delegated to by `Toolkit.serve`. When `a2a` is absent,
 * the server answers 404 to everything (a minimal base for now).
 */
export async function startA2AServer(opts: StartServerOptions): Promise<ServeHandle> {
  const { a2a, skills, runTask, onTask, mcp, mcpTools, onCall } = opts
  const [host, portStr] = splitAddr(opts.addr)
  const store = a2a ? resolveStore(a2a.store) : undefined
  const exposedTools = mcp ? exposedMcpTools(mcpTools ?? [], mcp) : []

  const server = http.createServer((req, res) => {
    const rpcRes = (id: unknown, payload: object) => {
      res.writeHead(200, { "content-type": "application/json" })
      res.end(JSON.stringify({ jsonrpc: "2.0", id: id ?? null, ...payload }))
    }
    const rpcOk = (id: unknown, result: unknown) => rpcRes(id, { result })
    const rpcErr = (id: unknown, code: number, message: string) => rpcRes(id, { error: { code, message } })

    // No profile at all ⇒ no routes.
    if (!store && !mcp) {
      res.writeHead(404)
      res.end("not found")
      return
    }

    // MCP streamable-HTTP profile (independent of A2A): a fresh server+transport
    // per request (stateless — sessionIdGenerator undefined). Checked first so
    // `/mcp` is never shadowed by the A2A POST handler.
    if (mcp && req.url && (req.url === "/mcp" || req.url.startsWith("/mcp?"))) {
      void (async () => {
        const body = req.method === "POST" ? await readBody(req) : ""
        const parsed = body ? safeJson(body) : undefined
        const mcpServer = buildMcpServer(exposedTools, mcp, onCall)
        const transport = new StreamableHTTPServerTransport({ sessionIdGenerator: undefined })
        res.on("close", () => {
          void transport.close()
          void mcpServer.close()
        })
        await mcpServer.connect(transport)
        await transport.handleRequest(req, res, parsed)
      })().catch(() => {
        if (!res.headersSent) {
          res.writeHead(500)
          res.end("mcp error")
        }
      })
      return
    }

    // No A2A profile ⇒ no A2A routes (an MCP-only server 404s everything else).
    if (!a2a || !store) {
      res.writeHead(404)
      res.end("not found")
      return
    }

    if (req.method === "GET" && req.url === CARD_PATH) {
      const url = baseUrl(host, server) + "/"
      res.writeHead(200, { "content-type": "application/json" })
      res.end(JSON.stringify(buildAgentCard(a2a, skills, url)))
      return
    }

    if (req.method === "POST") {
      readBody(req).then((body) => {
        let rpc: any
        try {
          rpc = JSON.parse(body)
        } catch {
          return rpcErr(null, -32700, "Parse error")
        }
        if (rpc?.method === "SendMessage") {
          const text = messageText(rpc.params)
          const contextId = typeof rpc.params?.message?.contextId === "string" ? rpc.params.message.contextId : undefined
          const id = randomUUID()
          const submitted: A2ATask = { id, contextId, status: { state: "submitted" } }
          // Persist submitted, kick fulfilment async, return the id immediately.
          store.save(submitted).then(() => {
            void fulfil(id, text, store, runTask, onTask, contextId)
            rpcOk(rpc.id, submitted)
          })
          return
        }
        if (rpc?.method === "GetTask") {
          const id = String(rpc.params?.id ?? "")
          store.get(id).then((task) => {
            if (!task) return rpcErr(rpc.id, -32001, `Task not found: ${id}`)
            rpcOk(rpc.id, task)
          })
          return
        }
        return rpcErr(rpc?.id ?? null, -32601, `Method not found: ${rpc?.method}`)
      })
      return
    }

    res.writeHead(404)
    res.end("not found")
  })

  await new Promise<void>((resolve) => server.listen(Number(portStr), host, resolve))

  return {
    url: baseUrl(host, server),
    stop() {
      return new Promise<void>((resolve) => server.close(() => resolve()))
    },
    close() {
      return new Promise<void>((resolve) => server.close(() => resolve()))
    },
  }
}

/**
 * Background fulfilment: submitted → working → (completed | failed). Never
 * throws — a fulfilment error becomes a `failed` Task, keeping the server alive.
 */
async function fulfil(
  id: string,
  text: string,
  store: TaskStore,
  runTask: (text: string, contextId?: string) => Promise<RunResult>,
  onTask?: OnTask,
  contextId?: string,
): Promise<void> {
  let task: A2ATask
  let result: RunResult | undefined
  let state: A2ATaskState
  try {
    await store.save({ id, status: { state: "working" } })
    // contextId groups a peer's turns into one conversation — thread it so the
    // client's ConversationStore remembers across tasks in the same context.
    result = await runTask(text, contextId)
    if (result.status === "pending" && result.pending) {
      // §10 suspension over A2A: the run halted waiting on out-of-band input. Surface the
      // protocol's `input-required` state carrying the request prompt — never a false `completed`.
      task = {
        id,
        status: { state: "input-required", message: { role: "agent", parts: [{ kind: "text", text: result.pending.prompt }] } },
      }
      state = "input-required"
    } else {
      const artifact: A2AArtifact = {
        artifactId: randomUUID(),
        parts: [{ kind: "text", text: result.text }],
      }
      task = { id, status: { state: "completed" }, artifacts: [artifact] }
      state = "completed"
    }
  } catch (e) {
    const detail = e instanceof Error ? e.message : String(e)
    task = { id, status: { state: "failed", message: { role: "agent", parts: [{ kind: "text", text: detail }] } } }
    state = "failed"
  }
  try {
    await store.save(task)
  } catch {
    // A store write failure must not crash the server.
  }
  if (onTask) {
    try {
      await onTask({ id, task, result, state })
    } catch {
      // Host callback errors are isolated.
    }
  }
}

// ---------------------------------------------------------------------------
// Address helpers.
// ---------------------------------------------------------------------------

/** Split `host:port` on the last colon. Missing host ⇒ 127.0.0.1. */
function splitAddr(addr: string): [string, string] {
  const idx = addr.lastIndexOf(":")
  if (idx === -1) return ["127.0.0.1", addr]
  const host = addr.slice(0, idx) || "127.0.0.1"
  return [host, addr.slice(idx + 1)]
}

/** Base URL for a listening server; 0.0.0.0 is reported as 127.0.0.1 for callers. */
function baseUrl(host: string, server: http.Server): string {
  const addr = server.address()
  const port = typeof addr === "object" && addr ? addr.port : 0
  const h = host === "0.0.0.0" || host === "::" ? "127.0.0.1" : host
  return `http://${h}:${port}`
}
